query/src/main/java/org/infinispan/search/mapper/work/impl/SearchIndexerImpl.java

```diff
@@ -1,39 +1,112 @@
 import org.hibernate.search.engine.backend.work.execution.OperationSubmitter;
-import org.hibernate.search.engine.common.execution.spi.SimpleScheduledExecutor;
 import org.hibernate.search.mapper.pojo.model.spi.PojoRawTypeIdentifier;
 import org.hibernate.search.mapper.pojo.route.DocumentRoutesDescriptor;
 import org.hibernate.search.mapper.pojo.work.spi.PojoIndexer;
+import org.infinispan.commons.reactive.RxJavaInterop;
 import org.infinispan.commons.util.concurrent.CompletableFutures;
-import org.infinispan.query.concurrent.InfinispanIndexingExecutorProvider;
+import org.infinispan.query.backend.QueryInterceptor;
+import org.infinispan.query.logging.Log;
 import org.infinispan.search.mapper.mapping.EntityConverter;
 import org.infinispan.search.mapper.session.impl.InfinispanIndexedTypeContext;
 import org.infinispan.search.mapper.session.impl.InfinispanTypeContextProvider;
 import org.infinispan.search.mapper.work.SearchIndexer;
-import org.infinispan.util.concurrent.BlockingManager;
+import org.infinispan.util.logging.LogFactory;
+
+import io.reactivex.rxjava3.core.Flowable;
+import io.reactivex.rxjava3.disposables.Disposable;
+import io.reactivex.rxjava3.internal.functions.Functions;
+import io.reactivex.rxjava3.processors.FlowableProcessor;
+import io.reactivex.rxjava3.processors.UnicastProcessor;
 
 /**
  * @author Fabio Massimo Ercoli
  */
 public class SearchIndexerImpl implements SearchIndexer {
+   private static final Log log = LogFactory.getLog(QueryInterceptor.class, Log.class);
 
    private final PojoIndexer delegate;
    private final EntityConverter entityConverter;
    private final InfinispanTypeContextProvider typeContextProvider;
 
-   private final SimpleScheduledExecutor offloadingExecutor;
+   private final int maxConcurrency;
+   private final ConcurrentMap<Supplier<Flowable<?>>, CompletableFuture<?>> submittedTasks;
+
+   private Disposable processorDisposer;
+
+   private FlowableProcessor<Supplier<Flowable<?>>> requestProcessor;
 
    public SearchIndexerImpl(PojoIndexer delegate, EntityConverter entityConverter,
-         InfinispanTypeContextProvider typeContextProvider, BlockingManager blockingManager) {
+                            InfinispanTypeContextProvider typeContextProvider, int maxConcurrency) {
       this.delegate = delegate;
       this.entityConverter = entityConverter;
       this.typeContextProvider = typeContextProvider;
-      this.offloadingExecutor = InfinispanIndexingExecutorProvider.writeExecutor(blockingManager);
+      this.maxConcurrency = maxConcurrency;
+
+      this.submittedTasks = new ConcurrentHashMap<>();
+   }
+
+   @Override
+   public void start() {
+      requestProcessor = UnicastProcessor.<Supplier<Flowable<?>>>create().toSerialized();
+      // onBackpressureLatest will drop any item that it can't immediately pass downstream
+      // Thus next downstream must batch them to not immediately drop
+      processorDisposer = requestProcessor.onBackpressureLatest(supplier -> {
```

Comment:
Could it cause some issues with the underlying index somehow? Say HS is not keeping up, and we start completing exceptionally and dropping. Later, it started working again, but some updates were dropped in the interval.

It is OK since we protect the application of OOMing, but could it cause some problem to the index itself (corruption or something similar)? If users receive the exception, they can't expect the entity to be indexed/up-to-date on the index.

Reply:
The indexing operation failing should prevent the write operation from occurring, so it should be consistent. The only case I don't know is if a putAll fails in the middle.

Reply:
The putAll case would require a tx of some sort to commit all the HS updates once the cache has been updated, but this is a consistency issue with putAll in general and not directly related to this PR.